plugins/UserId/Archiver.php

```diff
@@ -0,0 +1,130 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Plugins\UserId;
+use Piwik\Container\StaticContainer;
+use Piwik\DataArray;
+use Piwik\DataTable;
+use Piwik\Metrics;
+
+/**
+ * Archiver that aggregates metrics per user ID (user_id field).
+ */
+class Archiver extends \Piwik\Plugin\Archiver
+{
+    const USERID_ARCHIVE_RECORD = "UserId_archive_record";
+
+    const VISITOR_ID_FIELD = 'idvisitor';
+    const USER_ID_FIELD = 'user_id';
+
+    /**
+     * @var DataArray
+     */
+    protected $arrays;
+
+    /**
+     * Array to save visitor IDs for every user ID met during archiving process. We use it to
+     * fill metadata before actual inserting rows to DB.
+     * @var array
+     */
+    protected $visitorIdsUserIdsMap = array();
+
+    /**
+     * Archives data for a day period.
+     */
+    public function aggregateDayReport()
+    {
+        $this->arrays = new DataArray();
+        $this->aggregateUsers();
+        $this->insertDayReports();
+    }
+    /**
+     * Period archiving: simply sums up daily archives
+     */
+    public function aggregateMultipleReports()
+    {
+        $dataTableRecords = array(self::USERID_ARCHIVE_RECORD);
+        $columnsAggregationOperation = null;
+        $this->getProcessor()->aggregateDataTableRecords(
+            $dataTableRecords,
+            $maximumRowsInDataTableLevelZero = null,
```

Comment:
Here we should set a limit of eg 50k rows by default. To do this I suggest to create a new config entry similar to custom variables see https://github.com/piwik/piwik/blob/2.16.0/config/global.ini.php#L393-L395

eg `datatable_archiving_maximum_rows_userid_users = 50000`. 

This way it will be fast for all users by default, even when having more than 50k users as it other users will be grouped into one summary row "others". This is the Piwik way of doing it currently. Users that want to see all users can simply set it to a high value like 10mio and it will make sure that all users are shown in the report. 

Basically you will set similar to https://github.com/piwik/piwik/blob/2.16.0/plugins/CustomVariables/Archiver.php#L49 
`$maximumRowsInDataTableLevelZero = Config::getInstance()->General['datatable_archiving_maximum_rows_userid_users'];`

It's not an ideal solution for sure but that's how it is supposed to work right now.